File: EV-877/MARC.Everest/DataTypes/QSS.cs
/* 
 * Copyright 2008-2012 Mohawk College of Applied Arts and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you 
 * may not use this file except in compliance with the License. You may 
 * obtain a copy of the License at 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.

 * 
 * User: Justin Fyfe
 * Date: 09-22-2011
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using MARC.Everest.Attributes;
using System.Xml.Serialization;
using MARC.Everest.DataTypes.Interfaces;
using MARC.Everest.Interfaces;

namespace MARC.Everest.DataTypes
{
    /// <summary>
    /// Represents a <see cref="T:QSET{T}"/> that has been specialized as containing a set.
    /// </summary>
    /// <seealso cref="T:QSET{T}"/>
    /// <seealso cref="T:QSD{T}"/>
    /// <seealso cref="T:QSI{T}"/>
    /// <seealso cref="T:QSP{T}"/>
    /// <seealso cref="T:QSS{T}"/>
    /// <seealso cref="T:QSU{T}"/>
    /// <seealso cref="T:SXPR{T}"/>
    /// <seealso cref="T:SXCM{T}"/>
    /// <seealso cref="T:GTS"/>
    [Serializable]
    [Structure(Name = "QSS", StructureType = StructureAttribute.StructureAttributeType.DataType)]
    [XmlType("QSS", Namespace = "urn:hl7-org:v3")]
    public class QSS<T> : QSET<T>, ICollection<T>, IEquatable<QSS<T>>, IListContainer
        where T : IAny
    {

        /// <summary>
        /// Creates a new instance of the QSET set class
        /// </summary>
        public QSS() : base() {
            this.Terms = new List<T>();
        }

        /// <summary>
        /// Creates a new instance of the QSET set class containing the specified <paramref name="collection"/>
        /// </summary>
        public QSS(List<T> terms)
        {
            this.Terms = terms;
        }

        /// <summary>
        /// Creates a new instance of the QSET set class containing the specified <paramref name="collection"/>
        /// </summary>
        public QSS(params T[] terms)
        {
            this.Terms = new List<T>(terms);
        }

        /// <summary>
        /// Equivalent set operator
        /// </summary>
        protected override SetOperator GetEquivalentSetOperator()
        {
            return SetOperator.Inclusive;
        }

        /// <summary>
        /// Gets or sets the terms that make up the union of this QSET
        /// </summary>
        [Property(Name = "term", Conformance = PropertyAttribute.AttributeConformanceType.Mandatory, PropertyType = PropertyAttribute.AttributeAttributeType.NonStructural)]
        [XmlElement("term")]
        public List<T> Terms { get; set; }

        /// <summary>
        /// Validate that the QSET is valid. A QSET is valid when it contains at least two sets 
        /// in the <see cref="P:Terms"/> property and contains no-null terms.
        /// </summary>
        public override bool Validate()
        {
            bool isValid = (this.NullFlavor != null) ^ (this.Terms != null && this.Terms.Count > 0);
            foreach (var qs in Terms ?? new List<T>())
                isValid &= qs != null && !qs.IsNull;
            return isValid;
        }


        #region IEnumerable<T> Members

        /// <summary>
        /// Get the enumerator for this QSET SET
        /// </summary>
        /// <exception cref="T:System.NullReferenceException">When the value of <see cref="P:Terms"/> has not been set</exception>
        public IEnumerator<T> GetEnumerator()
        {
            return this.Terms.GetEnumerator();
        }

        #endregion

        #region IEnumerable Members

        /// <summary>
        /// Get the non-generic enumerator
        /// </summary>
        /// <exception cref="T:System.NullReferenceException">When the value of <see cref="P:Terms"/> has not been set</exception>
        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        #endregion

        #region ICollection<T> Members

        /// <summary>
        /// Add an expression to the term set
        /// </summary>
        public void Add(T item)
        {
            this.Terms.Add(item);
        }

        /// <summary>
        /// Clear the terms
        /// </summary>
        public void Clear()
        {
            this.Terms.Clear();
        }

        /// <summary>
        /// Determine if the terms contain the item
        /// </summary>
        public bool Contains(T item)
        {
            return this.Terms.Contains(item);
        }

        /// <summary>
        /// Copy the terms to an array
        /// </summary>
        public void CopyTo(T[] array, int arrayIndex)
        {
            this.Terms.CopyTo(array, arrayIndex);
        }

        /// <summary>
        /// Count the items in the terms
        /// </summary>
        public int Count
        {
            get { return this.Terms.Count; }
        }

        /// <summary>
        /// True if the terms are readonly
        /// </summary>
        public bool IsReadOnly
        {
            get { return false; }
        }

        /// <summary>
        /// Remove a term from the collection
        /// </summary>
        public bool Remove(T item)
        {
            return this.Terms.Remove(item);
        }

        #endregion

        #region IEquatable<QSS<T>> Members

        /// <summary>
        /// Determine equality between two QSS instances
        /// </summary>
        public bool Equals(QSS<T> other)
        {
            bool result = base.Equals(other);
            result &= this.Terms.Count == other.Terms.Count;
            for (int i = 0; i < other.Terms.Count && result; i++)
                result &= this.Terms[i].Equals(other.Terms[i]);
            return result;
        }

        /// <summary>
        /// Determine equality between two items
        /// </summary>
        public override bool Equals(object obj)
        {
            if (obj is QSS<T>)
                return Equals((QSS<T>)obj);
            return base.Equals(obj);
        }
        #endregion

        #region IListContainer Members

        /// <summary>
        /// Gets or sets the contents of the list contained by the object
        /// </summary>
        LIST<MARC.Everest.Interfaces.IGraphable> IListContainer.ContainedList
        {
            get
            {
                return new LIST<IGraphable>(this.Terms);
            }
            set
            {
                this.Terms = new List<T>(LIST<T>.Parse(value));
            }
        }

        #endregion

        /// <summary>
        /// No normalization needed
        /// </summary>
        public override IGraphable Normalize()
        {
            return this.Clone() as IGraphable;   
        }
    }
}

File: jEverest/src/ca/marc/everest/datatypes/generic/QSS.java
/* 
 * Copyright 2008-2012 Mohawk College of Applied Arts and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you 
 * may not use this file except in compliance with the License. You may 
 * obtain a copy of the License at 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.
 * 
 * User: Justin Fyfe
 * Date: 10-11-2012
 */
package ca.marc.everest.datatypes.generic;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

import ca.marc.everest.annotations.*;
import ca.marc.everest.datatypes.EverestValidationMessages;
import ca.marc.everest.datatypes.SetOperator;
import ca.marc.everest.datatypes.interfaces.IAny;
import ca.marc.everest.datatypes.interfaces.IListContainer;
import ca.marc.everest.datatypes.interfaces.ISetComponent;
import ca.marc.everest.interfaces.IGraphable;
import ca.marc.everest.interfaces.IResultDetail;
import ca.marc.everest.interfaces.ResultDetailType;
import ca.marc.everest.resultdetails.DatatypeValidationResultDetail;

/**
 * Represents a QSET{T} that has been specialized to contain a list of primitive values which
 * are unioned to provide a complete set of data
 */
@Structure(name = "QSS", structureType = StructureType.DATATYPE)
public class QSS<T extends IAny> extends QSET<T> implements IListContainer, Collection<T> {

	// The list of terms
	private List<T> m_terms = new ArrayList<T>();
	
	/**
	 * Creates a new instance of QSS
	 */
	public QSS() { super(); }
	/**
	 * Creates a new instance of QSS with the specified collection of terms
	 */
	public QSS(Collection<T> terms)
	{
		this();
		this.m_terms = new ArrayList<T>(terms);
	}
	/**
	 * Creates an instance of QSS with the specified terms
	 */
	public static <T extends IAny> QSS<T> createQSS(T... terms)
	{
		return new QSS<T>(Arrays.asList(terms));
	}

	/**
	 * Gets the value which represents the terms
	 */
	@Property(name = "term", propertyType = PropertyType.NONSTRUCTURAL, conformance = ConformanceType.REQUIRED)
	public List<T> getTerms() { return this.m_terms; }
	/**
	 * Sets the terms in the collection
	 */
	public void setTerms(List<T> value) { this.m_terms = value; }
	
	/**
	 * Normalize the items
	 */
	@SuppressWarnings("unchecked")
	@Override
	public IGraphable normalize() {
		QSS<T> retVal = (QSS<T>)this.shallowCopy();
		return retVal;
	}

	/**
	 * Get the list contained in this collection
	 */
	@Override
	public List<IGraphable> getContainedList() {
		return new LIST<IGraphable>(this.m_terms);
	}

	/**
	 * Set the contained list
	 */
	@Override
	public void setContainedList(List<IGraphable> value) {
		this.m_terms = new ArrayList<T>(new LIST<T>(value));
	}

	/**
	 * Get the equivalent set operator
	 */
	@Override
	protected SetOperator getEquivalentSetOperator() {
		return SetOperator.Inclusive;
	}
	/**
	 * Add an item to the collection
	 */
	@Override
	public boolean add(T term) {
		return this.m_terms.add(term);
	}
	/**
	 * Add all items from the collection to the list of terms
	 */
	@Override
	public boolean addAll(Collection<? extends T> terms) {
		return this.m_terms.addAll(terms);
	}
	/**
	 * Clears the item from the list
	 */
	@Override
	public void clear() {
		this.m_terms.clear();
	}
	/**
	 * Returns true if the collection contains the specified item
	 */
	@Override
	public boolean contains(Object term) {
		return this.m_terms.contains(term);
	}
	/**
	 * Returns true if the collection of terms contains all the 
	 * terms in the specified array.
	 */
	@Override
	public boolean containsAll(Collection<?> terms) {
		return this.m_terms.containsAll(terms);
	}
	/**
	 * Returns true if the collection is empty
	 */
	@Override
	public boolean isEmpty() {
		return this.m_terms != null && this.m_terms.size() == 0;
	}
	/**
	 * Gets the iterator for the collection
	 */
	@Override
	public Iterator<T> iterator() {
		return this.m_terms.iterator();
	}
	/**
	 * Remove an object from the list of terms
	 */
	@Override
	public boolean remove(Object term) {
		return this.m_terms.remove(term);
	}
	/**
	 * Remove all the items from the list of terms
	 */
	@Override
	public boolean removeAll(Collection<?> terms) {
		return this.m_terms.removeAll(terms);
	}
	/**
	 * Retain all items in the collection
	 */
	@Override
	public boolean retainAll(Collection<?> terms) {
		return this.m_terms.retainAll(terms);
	}
	/**
	 * Gets the size of the collection
	 */
	@Override
	public int size() {
		return this.m_terms.size();
	}
	/**
	 * Conver tthis collection to an array
	 */
	@Override
	public Object[] toArray() {
		return this.m_terms.toArray();
	}
	/**
	 * Convert this collection to an array
	 */
	@Override
	public <T> T[] toArray(T[] typeColl) {
		return this.m_terms.toArray(typeColl);
	}
	/**
	 * Validate extended
	 */
	@Override
	public Collection<IResultDetail> validateEx() {
		List<IResultDetail> retVal = new ArrayList<IResultDetail>();

        if (!(this.isNull() ^ !this.isEmpty()))
            retVal.add(new DatatypeValidationResultDetail(ResultDetailType.ERROR, "QSS", EverestValidationMessages.MSG_NULLFLAVOR_WITH_VALUE, null));
        for (T qs : this)
            if (qs == null || qs.isNull())
                retVal.add(new DatatypeValidationResultDetail(ResultDetailType.ERROR, "QSS", EverestValidationMessages.MSG_NULL_COLLECTION_VALUE, null));
        return retVal;
	}
	/**
	 * Validate the QSS instance
	 */
	@Override
	public boolean validate() {
		boolean isValid = this.isNull() ^ !this.isEmpty();
        for (T qs : this)
            isValid &= qs != null && !qs.isNull();
        return isValid;
	}
	/** (non-Javadoc)
	 * @see java.lang.Object#hashCode()
	 */
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result + ((m_terms == null) ? 0 : m_terms.hashCode());
		return result;
	}
	/** (non-Javadoc)
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (!super.equals(obj))
			return false;
		if (getClass() != obj.getClass())
			return false;
		QSS other = (QSS) obj;
		if (m_terms == null) {
			if (other.m_terms != null)
				return false;
		} else if (!m_terms.equals(other.m_terms))
			return false;
		return true;
	}

}
